src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs

```diff
@@ -0,0 +1,293 @@
+            // (Relative to 0 or the previous token’s start if they're on the same line)
+            var startCharacterDelta = startCharacter;
+            if (lastLineNumber == lineNumber)
+            {
+                startCharacterDelta = startCharacter - lastStartCharacter;
+            }
+
+            // 3. Token length
+            var tokenLength = textSpan.Length;
+
+            // Getting the classified spans that are modifiers.
+            // Since a TextSpan can have multiple ClassifiedSpans, all of the ClassifiedSpans except one (the token)
+            // should be a modifier.
+            var modifiers = textSpanToClassifiedSpans.Where(
+                s => ClassificationTypeNames.AdditiveTypeNames.Contains(s.ClassificationType));
+
+            // Filtering out the modifiers for now (will be added in step 5). We just want the primary token.
+            var tokenTypeClassifiedSpan = textSpanToClassifiedSpans.Except(modifiers).Single();
+
+            // 4. Token type - looked up in SemanticTokensLegend.tokenTypes (language server defined mapping
+            // from integer to LSP token types).
+            if (!s_classificationTypeToSemanticTokenTypeMap.TryGetValue(tokenTypeClassifiedSpan.ClassificationType, out var tokenTypeStr))
+            {
+                throw new NotSupportedException($"Classification type {tokenTypeClassifiedSpan.ClassificationType} is unsupported.");
+            }
+
+            var tokenTypeIndex = GetTokenTypeIndex(tokenTypeStr);
+
+            // 5. Token modifiers - each set bit will be looked up in SemanticTokensLegend.tokenModifiers
+            var modifierBits = TokenModifiers.None;
+            foreach (var currentModifier in modifiers)
+            {
+                if (!s_classificationTypeToSemanticTokenModifierMap.TryGetValue(currentModifier.ClassificationType, out var modifier))
+                {
+                    throw new NotSupportedException($"Classification type {currentModifier.ClassificationType} is unsupported.");
+                }
+
+                modifierBits |= modifier;
+            }
+
+            lastLineNumber = lineNumber;
+            lastStartCharacter = startCharacter;
+
+            deltaLineOut = deltaLine;
+            startCharacterDeltaOut = startCharacterDelta;
+            tokenLengthOut = tokenLength;
+            tokenTypeOut = tokenTypeIndex;
+            tokenModifiersOut = (int)modifierBits;
+        }
+
+        // Note: Method is internal since it is used by tests
+        internal static int GetTokenTypeIndex(string? tokenTypeStr)
+        {
+            var tokenTypeIndex = 0;
+            foreach (var tokenType in LSP.SemanticTokenTypes.AllTypes)
+            {
+                if (tokenType == tokenTypeStr)
+                {
+                    break;
+                }
+
+                tokenTypeIndex += 1;
+            }
+
+            return tokenTypeIndex;
+        }
```

Comment:
this seems bad.  this is N^2 as we can have N tokesn that each do a N lookup here.   Seems like we can compute and cache the map from `.ClassificationType` to `index`.